### PR TITLE
Multiple Home Workspaces: Move EngineController, Scheduler to HomeWorkspaceModel

### PR DESCRIPTION
**Pull request by pboyer**
### Purpose

This pull request makes it possible for a user to have more than one open home workspace at a time.
### Impact assessment
- [ ] Changes the way 3rd party nodes work (Add ikeough to the reviewer list)
- [ ] Includes migrations (Add ikeough to the reviewer list)
- [x] Changes a public interface (Add lukechurch to the reviewer list)
- [x] Affects which components of the system depends on which other components (Add pboyer to the reviewer list)
- [x] Introduces new components (Add lukechurch to the reviewer list)
- [x] Has a performance impact (Add keyu to the reviewer list)
- [ ] Changes the installer (Add sharadkjaiswal to the reviewer list)
- [ ] User facing strings, if any, are extracted into `*.resx` files
#### References

TBD
- Link to communications doc (Needed for changes that impact public API)
#### Testing strategy

All existing tests should succeed. New tests should be added to validate the work.
### Review staging
- [ ] Architectural Design Review - I discussed the design with @ke-yu, @Benglin, and @lukechurch 
- [ ] UI Review - There are essentially no UI changes here.
### Reviewers

TBD

### CODE REVIEW

**Comment by ke-yu** [comment]
awesome！

**Comment by pboyer** [comment]
@ke-yu It would be great to get your feedback on this - can I make you co-reviewer? A lot of this work is not fun (e.g. `DynamoViewModel.IsHomeSpace`). However, there's a few points I'm not so familiar with. For example, I'm proposing to simply have multiple `EngineController`'s, one on each `HomeWorkspaceModel`. It already works, but does that make sense?

**Comment by ke-yu** [comment]
@pboyer , yes it makes sense, each `HomeWorkspaceModel` should have its own `EngineController`. @junmendoza , correct me if I'm wrong -- now LiveRunner runs on main thread, right? We could have a separate thread to run graph so that multiple homeworkspace could run in parallel.

**Comment by pboyer** [comment]
@ke-yu Thank you kindly for your reply! Yes, this work is currently under development, unlike my previous pull request which was purely experimental. Our reason to prioritize this was largely practical - the longer we prolong not adding multiple "home" workspaces, the more developers write code assuming it. :)

Threading certainly is of utmost concern. As I understood it, the scheduler enforces Dynamo's threading contracts and would need to be updated in order to allow for multiple, concurrent runs. This would only be allowed in Standalone/Studio, of course. In Revit, the requirement is that all runs take place on the UI thread, preventing concurrent runs from being initiated.

@Benglin Perhaps, you can also speak to the sensitive issues I may encounter here?

Another question for everyone, it seems like it would make more sense to change the name "Home Workspace" - I was thinking simply "Project". My reasoning is as follows:

1. It is strange to have multiple "home" workspaces. Most people only have one "home" (excluding the very wealthy), so this would be closer to user expectations. :)
2. The term "Project" is already familiar to Revit users and also happens to be nicely consistent with this user interaction. Where a Revit user brings parametric families into their project (rvt file), a Dynamo user brings parametric custom nodes into their home workspace. Similarly, like families can be nested, so can custom nodes.
3. The above parallelism also extends nicely to the recursive structure of custom nodes - families are similarly recursive.
4. In Revit, the term project refers to the primary "end state" that is to be executed as a building. Where Revit has "regeneration", Dynamo has DesignScript's associative evaluation.

**Comment by jnealb** [comment]
Yes on all counts

**Comment by Racel** [comment]
@pboyer- Good points. I also agree. I have updated the Home Workspace UX mockups. 

All- please take a look at the mockups. The proposed updates can be found in [MAGN-8288](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8288). I have reassigned this task to @pboyer for review for now. Let me know your thoughts!

**Comment by Racel** [comment]
@pboyer - so, I was just thinking about why 'Project' might not be the ideal replacement for 'Home workspace'.  Since the term 'Project' is already used in Revit, a Dynamo for Revit user could possibly have multiple Revit projects and Dynamo projects open simultaneously. By default, Revit projects are  named 'Project#'. If we do the same thing, it could cause some confusion as users are working with files and communicating about those files with others. What about we just change 'Home workspace' to simply 'Workspace' instead? Thoughts? I can update the mockups as needed.

**Comment by pboyer** [comment]
@Racel I'm not sure. Would it be like:

New Workspace...
New Custom Node...

That might work, but keep in mind that we already think about the workspace as the host for nodes and connectors. I wonder what @ikeough thinks...

**Comment by jnealb** [comment]
@Racel but as long as it is clear that you are working in a Dynamo project (ProjectD?) might there not be some advantage to having Dynamo and Revit "Projects" linked, as eventually they may be tied together even more than they are now? When you open a new Dynamo project, perhaps you should be presented with the option of inheriting the Revit Project# for your Dynamo Project naming (as a low-tech means to make the link); this could be a nice feature if some design went into it, no? I often wish for something like this, instead of hunting around for the correct rvt file etc when my Dyn gives me no help in this regard.

**Comment by ikeough** [comment]
@pboyer @Racel A short tour through Dynamo history... Workspace used to be Workbench (I was really into skeumorphism). Workspace came about at some point when we converted the internal `dynWorkbench` class into `WorkspaceModel`. 

Workspace is a somewhat arbitrary term. The way I see it, we have three options:
1. Stay with "Workspace". We've been using it this long. Users are familiar with it, and it's no more or less generic than "Project". Also, the internal classes we use are referred to as `HomeWorkspaceModel`, etc. It'll be difficult for developers to switch gears and talk to users about "Projects" when internally they'll be thinking "WorkspaceModels". Plus, Dynamo is beginning to own this term.
2. Switch to "Project". Project seems just as all-encompassing as "Workspace". Users will take a little convincing. And, it's not Dynamo-specific. But, it's familiar. It's easier to say "my Dynamo Project", then "my Dynamo Workspace".
3. Get really specific, and change the name to something like "Graph". This is how many users and developers refer to the workspace already. There's not dislocation between Graph and nodes and connectors, like there is between Workspace and nodes and connectors. The problem with Graph is that I'm not sure Dynamo will always have nodes and connectors. That's the representational style we use now, but it could be something different in the future, and it's certainly something different for Reach.

**Comment by pboyer** [comment]
@ikeough I think you might have gotten a confused message. I'm only proposing one thing - that we change "Home" workspace to "Project" workspace. My reasoning is found above.

**Comment by ikeough** [comment]
@pboyer Above your said:
>  " I've proposed to @Racel that we change the name "Home Workspace" - I was thinking simply "Project"."  

I took this to mean that you wanted to replace "Home Workspace" with "Project". 

Are you proposing that "Home Workspace" be called "Project Workspace" instead?

**Comment by pboyer** [comment]
Yup!

File: src/DynamoCore/Models/WorkspaceModel.cs [hunk 1]
```diff
@@ -720,17 +720,16 @@ public virtual void Clear()
         ///     If successful, the CurrentWorkspace.FilePath field is updated as a side effect
         /// </summary>
         /// <param name="newPath">The path to save to</param>
-        /// <param name="core"></param>
         /// <param name="isBackup">Indicates whether saved workspace is backup or not. If it's not backup,
         /// we should add it to recent files. Otherwise leave it.</param>
-        public virtual bool SaveAs(string newPath, ProtoCore.RuntimeCore runtimeCore, bool isBackup = false)
```
Comment (pboyer) [inline src/DynamoCore/Models/WorkspaceModel.cs]: This method required a `RuntimeCore` due to trace erroneously living on `HomeWorkspaceModel`. I generalized this method (refactoring `SerializeSessionData` to `SaveCustomData`) and moved the implementation having to do with trace to where it belongs - `HomeWorkspaceModel`. Note that `DynamoModel` does not have a `RuntimeCore`, it lives on the `HomeWorkspaceModel`.

File: src/DynamoCoreWpf/Commands/AutomationSettings.cs [hunk 2]
```diff
@@ -116,33 +116,6 @@ internal bool CanSaveRecordedCommands
...
-        /// <returns>Returns true if there is a pending evaluation and that the 
-        /// shutdown process should be deferred.</returns>
-        /// 
-        private bool HasPendingEvaluation
-        {
-            get
-            {
-                return
-                    owningDynamoModel.Workspaces.OfType<HomeWorkspaceModel>()
-                        .Any(hw => hw.IsEvaluationPending);
```
Comment (pboyer) [inline src/DynamoCoreWpf/Commands/AutomationSettings.cs]: `HomeWorkspaceModel` simply returned a literal `false` for `IsEvaluationPending` - this simply removes this useless method.

**Comment by Racel** [comment]
@pboyer @ikeough @riteshchandawar @lillismith - I updated the mockups in MAGN-8288. I vote for keeping 'Workspace' instead of switching to 'Project' for the reasons I stated above (avoid naming confusion with Revit Projects).

File: src/DynamoCore/Models/DynamoModel.cs [hunk 3]
```diff
@@ -702,52 +759,14 @@ private void LibraryLoaded(object sender, LibraryServices.LibraryLoadedEventArgs
...
-                            executionTimeSpan);
-
-                        Debug.WriteLine(String.Format(Resources.EvaluationCompleted, executionTimeSpan));
-
-                        ExecutionEvents.OnGraphPostExecution();
-                    }
-                    break;
-            }
-        }
-
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: All of these methods are moved into `OnEvaluationStarted` and `OnEvaluationCompleted`.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 4]
```diff
@@ -622,69 +595,6 @@ private void RemoveExtension(IExtension ext)
                 logSource.MessageLogged -= LogMessage;
         }
 
-        private void EngineController_TraceReconcliationComplete(TraceReconciliationEventArgs obj)
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: This code has been moved to `HomeWorkspaceModel`.

File: src/DynamoCore/Models/HomeWorkspaceModel.cs [hunk 5]
```diff
@@ -548,6 +591,108 @@ private void OnPreviewGraphCompleted(AsyncTask asyncTask)
        
         #endregion
 
+        #region Trace 
+
+        /// <summary>
+        /// Add trace data to the document before being saved
+        /// </summary>
+        /// <param name="document">The XmlDocument being saved</param>
+        protected override void SaveCustomData(XmlDocument document)
```
Comment (pboyer) [inline src/DynamoCore/Models/HomeWorkspaceModel.cs]: This code is generalized to store more than just `SessionData`.

**Comment by pboyer** [comment]
Ooohhhh

**Comment by ikeough** [comment]
@pboyer This is so stale it's going to be hard to update. But I see it as a necessary pre-requisite for the work to unify workspace models. From the initial merge I see almost the entirety of `HomeWorkspaceModel` is in conflict. Other files are not so bad. There may be a better strategy.

**Comment by pboyer** [comment]
I think there might be more than one PR. I need to get into the office to
ensure this is the latest.

**Comment by ikeough** [comment]
I've closed this as I have a more up to date branch with this work in it that we should use going forward.

**Pull request closed without merging**